### PR TITLE
dblatex: Support external TeXLive (in addition to MacTeX).

### PR DESCRIPTION
**Pull request by lemzwerg**
#### Description

This is a first try adding external TeXLive support in addition to MacTeX.  Both cases are handled by the `+mactex` variant for simplicity.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.7.5 11G63
Xcode 4.6.3 4H1503 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [ ] tried a full install with `sudo port -vst install`? → only `-vs` works; option `-t` doesn't work with external TeXLive
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by lemzwerg** [comment]
What shall I do now?  Are these failures related at all to my changes?

**Comment by mojca** [comment]
It says that the build timed out. This is not a failure, just lack of confirmation whether it builds or not. You cannot do anything about it (other than help us write a safe way to test PRs elsewhere), it's hardcoded time limit of Travis.

**Comment by pmetzger** [comment]
@lemzwerg Any time a build would take more than ~48 minutes it times out, and the builds at Travis don't take place on fast dedicated hardware, so for large packages or large updates this does indeed happen relatively routinely. (Indeed, Apple no longer makes rack mount machines, and the Mac Mini, which used to be used a lot as a substitute in such applications, has not been updated in some years.)

**Comment by lemzwerg** [comment]
Thanks for the info.  I wonder how the Homebrew people solve such situations – basically, they have exactly the same problems, haven't they?

**Comment by pmetzger** [comment]
I don't know. One has two choices: one can use something other than Travis that will run one's CI job no matter how long it takes, or one can use Travis. I know some projects that use both Travis and internal CI machines. We are unfortunately constrained by the fact that Apple allows Apple OSes to run only on Apple hardware, so we can't just spin up boxes at AWS or what have you.

**Comment by lemzwerg** [comment]
Hmm.  How shall I proceed so that my pull requests get reviewed and accepted?  I'm rather sure that my forthcoming improvements to `lilypond-devel` (and maybe other packages that could benefit from an external TeXLive distribution) will produce time-outs too...

**Comment by mojca** [comment]
Passing build is not required to accept the PR.

File: textproc/dblatex/Portfile [hunk 1]
```diff
@@ -125,10 +122,9 @@ proc dblatex.mktexlsr {} {
     if {![variant_isset mactex]} {
         texlive.mktexlsr
     } else {
-        global dblatex.mactex_bin
-        system "${dblatex.mactex_bin}/mktexlsr"
-        if [file exists "${dblatex.mactex_bin}/mtxrun"] {
-            system "${dblatex.mactex_bin}/mtxrun --generate"
+        system "mktexlsr"
+        if {[llength [auto_execok mtxrun]]} {
+            system "mtxrun --generate"
```
Comment (mojca) [inline textproc/dblatex/Portfile]: This is not really related to your PR, but I'm pretty sure that running `mtxrun` here makes no sense. It has to be done when new fonts or other useful resources are added to the TeX tree. I see no use of `mtxrun` for `dblatex` though.

File: textproc/dblatex/Portfile [hunk 2]
```diff
@@ -125,10 +122,9 @@ proc dblatex.mktexlsr {} {
     if {![variant_isset mactex]} {
         texlive.mktexlsr
     } else {
-        global dblatex.mactex_bin
-        system "${dblatex.mactex_bin}/mktexlsr"
-        if [file exists "${dblatex.mactex_bin}/mtxrun"] {
-            system "${dblatex.mactex_bin}/mtxrun --generate"
+        system "mktexlsr"
```
Comment (mojca) [inline textproc/dblatex/Portfile]: I don't particularly like this since you don't know which `mktexlsr` will be run here (the one from MacPorts or the one from external distribution).

File: textproc/dblatex/Portfile [hunk 3]
```diff
@@ -55,42 +55,39 @@ destroot.destdir    --root=${destroot} \
...
     depends_lib-delete \
                     port:texlive-latex-extra \
                     port:texlive-latex-recommended \
                     port:texlive-math-science
 
-    notes "The mactex variant will install styles to MacTeX's texmf-local
-        ${dblatex.texmflocal}
-    (which is outside macport's common directory structure).
-    For the mactex variant to work,
-    ${dblatex.mactex_bin} must also be added to binpath in macports.conf"
```
Comment (mojca) [inline textproc/dblatex/Portfile]: Why did you delete this note? I suspect it's still needed.

File: textproc/dblatex/Portfile [hunk 4]
```diff
@@ -55,42 +55,39 @@ destroot.destdir    --root=${destroot} \
...
+    directory (for example /Library/TeX/texbin) must be added to binpath\
+    in macports.conf *before* installing this port.} {
+
+    # Check whether we have a TeXLive manager (which MacPorts doesn't
+    # provide).  We boldly assume that other basic TeXLive binaries
+    # like kpsewhich are then available also.
+    if {[llength [auto_execok tlmgr]]} {
         set dblatex.texmflocal \
-            [exec ${dblatex.mactex_bin}/kpsewhich --expand-var='\$TEXMFLOCAL']
+            [exec kpsewhich --expand-var='\$TEXMFLOCAL']
```
Comment (mojca) [inline textproc/dblatex/Portfile]: See the other comment.

File: textproc/dblatex/Portfile (hunk 3 above)
Comment (lemzwerg) [reply textproc/dblatex/Portfile]: ???  It's not removed but reworded, with improved formatting.

File: textproc/dblatex/Portfile (hunk 2 above)
Comment (lemzwerg) [reply textproc/dblatex/Portfile]: OK, will upload a different solution.

File: textproc/dblatex/Portfile (hunk 3 above)
Comment (mojca) [reply textproc/dblatex/Portfile]: The first part is reworded. OK, the second part was moved to `else return error ...` which probably makes sense. I saw it in variant description, but somehow missed the one in the error report.

File: textproc/dblatex/Portfile (hunk 2 above)
Comment (lemzwerg) [reply textproc/dblatex/Portfile]: I've re-introduced `dblatex.mactex_bin` (but deriving it from the path).

File: textproc/dblatex/Portfile (hunk 1 above)
Comment (lemzwerg) [reply textproc/dblatex/Portfile]: I've removed the call to `mtxrun`.

File: textproc/dblatex/Portfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+# vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
```
Comment (mojca) [inline textproc/dblatex/Portfile]: Is there any particular reason for changing the modeline (other than perhaps vim not respecting the setting)? If you want some changes in this line, it makes sense to ask on the macports-devel mailing list rather than deliberately changing a single `Portfile` to be inconsistent with the rest.

Comment (lemzwerg) [reply textproc/dblatex/Portfile]: Well, it avoids an overlong line, also separating vim and Emacs settings (which are thus easier to comprehend IMHO).  Additionally, there is at least one other Portfile (`python/py-cdb/Portfile`) that also uses an Emacs local variables block at the end of the file – I thus assume that it is an allowed (I dare to say, better) variant.  I *did* check :-)

Comment (mojca) [reply textproc/dblatex/Portfile]: Still, please bring this discussion up to the developer mailing list. I don't feel competent enough to judge about it. But I feel that we should strive to keep the design consistent, and either change all modelines to what is considered a better solution, or none.

**Comment by mojca** [comment]
Summary: I did not test the changes yet (and there are some parts of Tcl which I would need to learn & test to fully understand), but apart from the (unrelated) modeline changes the patch conceptually looks ok to me. (If we ever get to a point of fully reproducible builds, this variant should get a "this cannot be built reproducibly" flag, but I'm fine with that :)

Yes, ideally I would like to see some PortGroup handle those dirty details in the long run, but let's not make the task more complicated than it should be right now.

For the modeline changes I would prefer to get feedback from other members of the community first. Even if the others agree about the modeline, it could make sense to decouple those changes from the functional ones (and apply them to more ports than just one).

File: textproc/dblatex/Portfile (hunk 5 above)
Comment (pmetzger) [reply textproc/dblatex/Portfile]: The guide says we have a standard modeline now. It should be left as-is. The fact that one port is inconsistent doesn't matter.

**Comment by pmetzger** [comment]
I don't want the modeline stuff changed. It's much less maintainable, and we already have a consensus that seems unproblematic.

**Comment by lemzwerg** [comment]
OK, I've undone the mode line change.

**Comment by lemzwerg** [comment]
Sooo...  What's the next step?  Anything I can do?

File: textproc/dblatex/Portfile [hunk 6]
```diff
@@ -57,40 +57,49 @@ destroot.destdir    --root=${destroot} \
...
+    notes "\
+The mactex variant will install styles into your TeXLive's texmf-local
+tree at
 
-    # AND, since we're installing files outside macports' normal directories
+    ${dblatex.texmflocal}
+
+(which is outside MacPort's common directory structure)."
+
+    # AND, since we're installing files outside MacpPrts' normal
```
Comment (mojca) [inline textproc/dblatex/Portfile]: Small typo

**Comment by mojca** [comment]
@neverpanic: thank you very much for the feedback. Files are already being installed outside of MacPorts when the `+mactex` variant is being used. It's some kind of a gray zone. We certainly cannot guarantee reproducibility of any kind and issues are to be expected, so this is only advisable to be done by expert users. I'm not a huge fan of this approach either, but those 5+ GB make the exception somewhat excusable. There was a long thread on the mailing list (I certainly wrote a long one myself, but I sent it with the wrong email a couple of times.) This patch is only supposed to improve the situation for those who don't have the installation at precisely some hardcoded path. The problems I found above need to be addressed of course.

**Comment by lemzwerg** [comment]
It would be actually very simple to avoid installation of files outside of MacPorts for the 'mactex' variant: Simply set up a texmf directory tree at a suitable MacPorts place, then register this tree within mactex's `texmf.cnf` file (namely, after the user's local texmf tree but before the texlive tree itself).  However, such a change would be (a) backwards incompatible – all mactex users had to update –, and (b) users had to fiddle with two config files instead of a single one.

**Comment by mojca** [comment]
If update of base is needed, the patch might need to wait for a new release. Which changes exactly are needed? Are they in the 2.5 branch also, or just in master? In my opinion the Tcl code should not fail (throw exceptions or whatever happens) just because the user doesn't have any TeX install. It's one thing if you get an error report (that's ok), and another one if Tcl interpreter fails.

I thought that this should be possible to fix by introducing a couple of more steps (first check whether a binary exists and only ask for normalize/dirname if it exists), or some exception handling mechanism.

But I really like your idea about changing the way how files are installed. It would be slightly more work, but it would be somewhat safer in many aspects in the long run. As I already mentioned, it would be by far the best to have a PortGroup that would be easy to use in individual portfiles which would take care of proper external TeX distribution discovery, which would automatically run `mktexlsr` (and `mtxrun --generate` if needed) at the end, know at which location the files should end up etc. I would suggest adding a separate `texmf` tree just for ports which are not part of TeX Live, so that overlapping packages would not conflict with each other when external distribution sees another `texmf` tree.

**Comment by lemzwerg** [comment]
No, no, the tcl failures you are experiencing are not related at all to my changes; Macports itself has changed – I did `port -d sync` (which automatically did `git pull`) and got 10 identical failures (the same you are experiencing); after re-compiling and re-installing 'macports-base' everything works fine again.

**Comment by lemzwerg** [comment]
And yes, in the long run such a change to the handling of texmf trees would be nice and preferable.  Please remember, however, that I'm not a Mac user at all – someone with more Macports experience has to implement this.

**Comment by pmetzger** [comment]
@mojca I'm going to leave this to you to merge when you feel it's time.

**Comment by lemzwerg** [comment]
Oops!  I haven't carefully read your error report, sorry.  I can reproduce the issue and will fix it soon.

**Comment by lemzwerg** [comment]
@mojca Branch updated. Please repeat your tests!

**Comment by mojca** [comment]
@lemzwerg: your changes are still problematic.

For testing purposes please add
```
default_variants +mactex
```
to the port (don't actually commit this line, just use it for checking) and make sure that you don't have any other `kpsewhich` in `binpath` (`macports.conf`).

This is what I then get:
```
$ port info dblatex
Error: dblatex: Error executing mactex: Cannot find MacTeX or external TeXLive installation; aborting.  Have you added the path to your TeX distribution's binary directory to binpath in macports.conf?
Can't map the URL 'file://.' to a port description file ("Error evaluating variants").
Please verify that the directory and portfile syntax are correct.
To use the current port, you must be in a port's directory.

$ portindex
Creating port index in /path/to/macports-ports
Error: dblatex: Error executing mactex: Cannot find MacTeX or external TeXLive installation; aborting.  Have you added the path to your TeX distribution's binary directory to binpath in macports.conf?
Failed to parse file textproc/dblatex/Portfile: Error evaluating variants

Total number of ports parsed:	1 
Ports successfully parsed:	0 
Ports failed:			1 
Up-to-date ports skipped:	20688
```
While the initial error message is correct, the file should not **fail to parse** when TeX is absent.

**Comment by lemzwerg** [comment]
Sorry for the late response. 

Are aware that the problem you describe is already present in the original Portfile?  In other words, someone else already introduced this behaviour, not me :-)

It's not clear to me how this can be fixed at all.  Do you have an example that works as you expect?

**Comment by mojca** [comment]
No, I didn't test the existing port as much, sorry :)
I assume that @nortcele contributed the initial functionality (https://trac.macports.org/ticket/55106#no1)?

I should take some time to test further, but what exactly happens with
```tcl
set dblatex.mactex_bin [file dirname [file normalize ${dblatex.temp}]]
```
when `${dblatex.temp}` is empty?

**Comment by lemzwerg** [comment]
```
tclsh
% set dblatex.temp ""
% set dblatex.mactex_bin [file dirname [file normalize ${dblatex.temp}]]
.
% echo ${dblatex.mactex_bin}
.
```

**Comment by mojca** [comment]
I'm sorry for the delay. I took some time to experiment. Maybe something in the following sense?

```
PortSystem          1.0

name                testport
version             1.0

variant foo_bad description {failing variant} {
    set foo_temp {*}[auto_execok foobin]
    set foo_bin [file dirname [file normalize ${foo_temp}]]
    if {${foo_bin} eq "."} {
       return -code error "Cannot find foobin.\
           Have you added the path to binpath in macports.conf?"
    }
}

variant foo_good description {failing variant} {
    if {![catch {set foo_temp {*}[auto_execok foobin]}]} {
        set foo_bin [file dirname [file normalize ${foo_temp}]]
        # when do you get a dot?
        if {${foo_bin} eq "."} {
            pre-fetch {
                return -code error "Cannot find foobin.\
                    Have you added the path to binpath in macports.conf?"
            }
        }
    } else {
        pre-fetch {
            return -code error "Binary doesn't exist."
        }
    }
}

default_variants +foo_good
```

We should make sure that non-existent binaries don't generate errors when simply parsing the port with `portindex` (even if the variant is not the default one). Throwing an error in `pre-fetch` is OK.

Even if the current behaviour is faulty, we should attempt to fix it.

**Comment by lemzwerg** [comment]
OK, this should be fixed now.  Using `catch` doesn't behave as I would expect – regardless whether  `set` works or doesn't work, the `if` clause gets called...  Due to my lack of Tcl knowledge I wasn't able to fix this, so I implemented it differently.

Note that you get useful information from
```
port notes dblatex +mactex
```
even if dblatex or an external TeXLive is not installed.

**Comment by jmroot** [comment]
> variant foo_good description {acceptable variant} {
>     if {![catch {set foo_temp {*}[auto_execok foobin]}]} {

Using {*} here doesn't make sense because `set` takes at most 2 arguments. If auto_execok returns a list, the code inside the catch will error.

Also it would be more idiomatic to write:
```
if {![catch {auto_execok foobin} result]} {
```
which will put the return value into `result` (regardless of whether there was an error).

**Comment by lemzwerg** [comment]
> Using {*} here doesn't make sense because set takes at most 2 arguments.

This is not correct.  Using `{*}` ensures that directories with spaces are properly flattened.  Here is a test, assuming the home directory is `/Users/Foo`:
```
cd ~
mkdir "This is a Test"
cd "This is a Test"
ln -s /opt/local/bin/port port
export PATH="/Users/Foo/This is a Test":$PATH

tclsh
% set foo [auto_execok port]
{/Users/Foo/This is a Test/port}
% file normalize $foo
/Users/Foo/This is a Test/{/Users/Foo/This is a Test/port}
% set bar {*}[auto_execok port]
/Users/Foo/This is a Test/port
% file normalize $bar
/Users/Foo/This is a Test/port
```

File: textproc/dblatex/Portfile [hunk 7]
```diff
@@ -56,41 +56,66 @@ destroot.destdir    --root=${destroot} \
...
         set dblatex.texmflocal \
             [exec ${dblatex.mactex_bin}/kpsewhich --expand-var='\$TEXMFLOCAL']
         regsub -all {'} ${dblatex.texmflocal} "" dblatex.texmflocal
+    }
+
+    # Make 'port notes dblatex +mactex' work even if neither the package
+    # nor an external TeXLive is installed.
+    if {${dblatex.texmflocal} eq ""} {
+        notes "The 'mactex' variant will install styles into your\
+            TeXLive's texmf-local tree (which is outside MacPort's\
```
Comment (ryandesign) [inline textproc/dblatex/Portfile]: Our name is "MacPorts" not "MacPort" so change `MacPort's` to either `MacPorts'` or `the MacPorts`.

File: textproc/dblatex/Portfile [hunk 8]
```diff
@@ -56,41 +56,66 @@ destroot.destdir    --root=${destroot} \
...
+            common directory structure)."
     } else {
-        return -code error "Cannot find MacTeX installation; aborting"
+        notes "\
+The 'mactex' variant will install styles into your TeXLive's texmf-local
```
Comment (ryandesign) [inline textproc/dblatex/Portfile]: Missing `\` at end of line.

File: textproc/dblatex/Portfile [hunk 9]
```diff
@@ -56,41 +56,66 @@ destroot.destdir    --root=${destroot} \
...
+tree at
+
+    ${dblatex.texmflocal}
+
+(which is outside MacPort's common directory structure)."
```
Comment (ryandesign) [inline textproc/dblatex/Portfile]: `MacPorts'` or `the MacPorts`.

**Comment by mojca** [comment]
@ryandesign: my main request would be reviewing the Tcl portion of the code which checks whether `kpsewhich` is available somewhere. You don't need to be a TeX expert for that (just a Tcl one :). Ideally one should start without having `kpsewhich` installed anywhere (or simply changing the name to something non-existent).

The idea is to check whether `kpsewhich` is available on the system (and making sure it's not the one installed by MacPorts, else this variant would make absolutely no sense). If it's not available, the port with that variant should refuse to install (but not fail to parse). If it's available, we need to run it to ask it where it wants the files.

Maybe even the current code does something that makes parsing the port alone fail in case `kpsewhich` is not installed anywhere.

We eventually need to put this code into a portgroup, but we can start with one properly working port for now ...

**Comment by ryandesign** [comment]
Sorry, I'm not going to have time to look into this further. There are a zillion other things I'm trying to work on.

**Comment by jmroot** [comment]
> > Using {*} here doesn't make sense because set takes at most 2 arguments.
> 
> This is not correct. Using `{*}` ensures that directories with spaces are properly flattened. Here is a test, assuming the home directory is `/Users/Foo`:
> 
> ```
> cd ~
> mkdir "This is a Test"
> cd "This is a Test"
> ln -s /opt/local/bin/port port
> export PATH="/Users/Foo/This is a Test":$PATH
> 
> tclsh
> % set foo [auto_execok port]
> {/Users/Foo/This is a Test/port}
> % file normalize $foo
> /Users/Foo/This is a Test/{/Users/Foo/This is a Test/port}
> % set bar {*}[auto_execok port]
> /Users/Foo/This is a Test/port
> % file normalize $bar
> /Users/Foo/This is a Test/port
> ```

And what if the list returned is of length 2 or more? Note I didn't say that using the return value of auto_execok directly was correct. Seems like lindex would be more appropriate.

**Comment by lemzwerg** [comment]
> And what if the list returned is of length 2 or more?

AFAIC, this can only happen on some MS Windows versions if you do what `auto_execok` was designed for, namely to get an argument list for a proper `exec` call.  On UNIX-like boxes, you always get one element.  In other words, this isn't an issue for MacPorts.

> Note I didn't say that using the return value of `auto_execok` directly was correct.
> Seems like `lindex` would be more appropriate.

Maybe.  Is it really worth the change?  IMHO, `{*}[...]` is easier to read than `[lindex [...] 0]`.

**Comment by jmroot** [comment]
> Is it really worth the change?

Probably not if the list will always be length 1.

**Comment by mojca** [comment]
Looks ok to me. Now the next step would to "portgroupify" these changes, so that one could use them in other ports. Thank you, @lemzwerg.
